clang/lib/AST/StmtProfile.cpp

```diff
@@ -1333,7 +1333,15 @@ void StmtProfiler::VisitPredefinedExpr(const PredefinedExpr *S) {
 void StmtProfiler::VisitIntegerLiteral(const IntegerLiteral *S) {
   VisitExpr(S);
   S->getValue().Profile(ID);
-  ID.AddInteger(S->getType()->castAs<BuiltinType>()->getKind());
+
+  int FoldingSetID = 0;
+
+  if (S->getType()->isBitIntType())
+    FoldingSetID = S->getValue().getSExtValue();
+  else
+    FoldingSetID = S->getType()->castAs<BuiltinType>()->getKind();
+
+  ID.AddInteger(FoldingSetID);
```

Comment:
I think that can lead to different expression having the same id.
Did you consider using `S->getBitWidthValue()` that gives you the number of bits in the bitfield + a boolean that tells you whether it is a bitfield?

Reply:
> and BitIntType has a Profile method

Neat, i did not check that

Reply:
Each `Type` subclass has a `Profile()` function that determines type identity. This is how we ensure that a `_BitInt(2)` and a `_BitInt(3)` can both be represented by `BitIntType` but not be the same type for things like redeclarations or overloading.

Reply:
Very minor nitpick: [`BuiltinType`](https://clang.llvm.org/doxygen/classclang_1_1BuiltinType.html) doesn't have a `Profile` method; the identity of builtin types is determined by the `getKind()` method (as it's used in this code fragment).

Reply:
Good point, I forgot that `BuiltinType` was a bit special.

Reply:
I believe I've addressed the comments in this thread. I'm resolving this conversation. Please reopen if you disagree.